### ISSUE TITLE
Support Certificate-Bound (POP) Opaque Access Token Validation

### ISSUE THREAD

**Issue by jgrandja** [spring-projects/spring-security#14888]
This issue is related to gh-10538 and will address validating certificate-bound "opaque" access tokens for both the Servlet and Reactive stacks.

**Comment by CrazyParanoid** [comment]
Hi @jgrandja ! It seems to me that such validation is impossible for opaque tokens, for example, in my case, the token is an ordinary guid. But even if we consider the rare case - jwt-based opaque token, then the introspection endpoint call must perform all the checks on its own and return the correct result. Currently `NimbusOpaqueTokenIntrospector` does not perform default validations, at most it does this:
```
        @Override
	public OAuth2AuthenticatedPrincipal introspect(String token) {
		RequestEntity<?> requestEntity = this.requestEntityConverter.convert(token);
		if (requestEntity == null) {
			throw new OAuth2IntrospectionException("requestEntityConverter returned a null entity");
		}
		ResponseEntity<String> responseEntity = makeRequest(requestEntity);
		HTTPResponse httpResponse = adaptToNimbusResponse(responseEntity);
		TokenIntrospectionResponse introspectionResponse = parseNimbusResponse(httpResponse);
		TokenIntrospectionSuccessResponse introspectionSuccessResponse = castToNimbusSuccess(introspectionResponse);
		// relying solely on the authorization server to validate this token (not checking
		// 'exp', for example)
		if (!introspectionSuccessResponse.isActive()) {
			this.logger.trace("Did not validate token since it is inactive");
			throw new BadOpaqueTokenException("Provided token isn't active");
		}
		return convertClaimsSet(introspectionSuccessResponse);
	}
```

**Comment by jgrandja** [comment]
@CrazyParanoid

> It seems to me that such validation is impossible for opaque tokens

Anything is possible. But currently there is no `OAuth2TokenValidator` associated with `OpaqueTokenAuthenticationProvider`, which is likely what is needed. This is my initial thought but that might change.

Either way, let's see how many upvotes this issue gets before we prioritize it in a release.

**Comment by jzheaux** [comment]
> It seems to me that such validation is impossible for opaque tokens

I think the API limitation is that `OpaqueTokenIntrospector` takes a string. This is because the token is opaque in the Introspection spec. Based on https://datatracker.ietf.org/doc/html/draft-ietf-oauth-pop-architecture-08#section-7.3, the certificate is verified directly by the resource server and not as part of the introspection handshake (See Figure 4).

In that case, it may be more appropriate to have the `OpaqueTokenAuthenticationProvider` validate the certificate, conditioned on a value in `BearerTokenAuthenticationToken` or a subclass.

**Comment by CrazyParanoid** [comment]
Maybe it would be better to make this task more global? For example, Support Opaque Token Validation. You can add an interface something like:
```
public interface OpaqueBearerTokenValidator<T extends BearerTokenAuthentication> {

	OAuth2TokenValidatorResult validate(T token);
}
```
Then in `OpaqueTokenAuthenticationProvider `:
```
        @Override
	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
		if (!(authentication instanceof BearerTokenAuthenticationToken bearer)) {
			return null;
		}
		OAuth2AuthenticatedPrincipal principal = getOAuth2AuthenticatedPrincipal(bearer);
		Authentication result = this.authenticationConverter.convert(bearer.getToken(), principal);
		if (result == null) {
			return null;
		}

		OAuth2TokenValidatorResult validationResult = validator.validate(result);
		if (validationResult.hasErrors()) {
			Collection<OAuth2Error> errors = validationResult.getErrors();
			String validationErrorString = getOpaqueTokenValidationMessage(errors);
			throw new OpaqueTokenValidationException(validationErrorString, errors);
		}

		if (AbstractAuthenticationToken.class.isAssignableFrom(result.getClass())) {
			final AbstractAuthenticationToken auth = (AbstractAuthenticationToken) result;
			if (auth.getDetails() == null) {
				auth.setDetails(bearer.getDetails());
			}
		}
		this.logger.debug("Authenticated token");
		return result;
	}
```
But even if this is implemented, it is necessary to make such validations optional, for example through setting `setOpaqueTokenValidation(true)`.

**Comment by jzheaux** [comment]
I don't think this would be appropriate in this case since such validation is not part of the introspection spec and thus not a part of the token. Further, the information that would be needed (the certificate itself) is neither part of the access token being introspected (it's just a `String` in this case) nor is it part of the Authorization Server response.

As per the diagram, the certificate is handed by the client to the resource server, separately from the token. The token is introspected, and the certificate is verified separately from that.

**Comment by CrazyParanoid** [comment]
I probably expressed myself incorrectly when I said that this is impossible. This is not practical. Since the authorization server performs all validation during the introspection, there is no point in implementing such validations in the framework.

**Comment by franticticktick** [comment]
Hi @jzheaux  and @jgrandja. I want to share my research with you: the authorization server can indeed return the `cnf` claim in the introspection response. Something like this: 
```
{
   "active":true,
   "scope":"write",
   "client_id":"myClient",
   "user_id":"myClient",
   "username":"myClient",
   "token_type":"Bearer",
   "exp":1547079953,
   "sub":"(age!myClient)",
   "subname":"myClient"
   "iss":"https://openam.example.com:8443/openam/oauth2",
   "cnf":{
      "x5t#S256":"m8UcWBSPNtaKN19TdR8zUHvWWOSCSX9nsa5vU6fscd0"
   }
}
```
And it seems that the certificate check can be done together with the response check. I did not find this in the RFC, but nevertheless many ID providers allow such response.
